File: client/python/memlinkclient.py
#coding: utf-8
import os, sys, types
from cmemlink import *

class MemLinkClietException (Exception):
    pass

class MemLinkClient:
    def __init__(self, host, readport, writeport, timeout):
        self.client = memlink_create(host, readport, writeport, timeout)
        if not self.client:
            raise MemLinkClietException

    def close(self):
        if self.client:
            memlink_close(self.client)

    def destroy(self):
        if self.client:
            memlink_destroy(self.client)
            self.client = None

    def ping(self):
        return memlink_cmd_ping(self.client)

    def dump(self):
        return memlink_cmd_dump(self.client)

    def clean(self, table, key):
        return memlink_cmd_clean(self.client, table, key)

    def create_table(self, table, valuesize, listtype, valuetype, attrstr=''):
        return memlink_cmd_create_table(self.client, table, valuesize, attrstr, listtype, valuetype)

    def create_table_list(self, table, valuesize, attrstr=''):
        return memlink_cmd_create_table_list(self.client, table, valuesize, attrstr);
    
    def create_table_queue(self, table, valuesize, attrstr=''):
        return memlink_cmd_create_table_queue(self.client, table, valuesize, attrstr);

    def create_table_sortlist(self, table, valuesize, valuetype, attrstr=''):
        return memlink_cmd_create_table_sortlist(self.client, table, valuesize, attrstr, valuetype);

    def create_node(self, table, key):
        return memlink_cmd_create_node(self.client, table, key)

    def stat(self, table, key):
        mstat = MemLinkStat()
        ret = memlink_cmd_stat(self.client, table, key, mstat)
        if ret != MEMLINK_OK:
            mstat = None
        return ret, mstat

    def stat_sys(self):
        mstat = MemLinkStatSys()
        ret = memlink_cmd_stat_sys(self.client, mstat)
        if ret != MEMLINK_OK:
            mstat = None
        return ret, mstat

    def delete(self, table, key, value):
        return memlink_cmd_del(self.client, table, key, value, len(value))

    def delete_by_attr(self, table, key, attr):
        return memlink_cmd_del_by_attr(self.client, table, key, attr)

    def insert(self, table, key, value, pos, attrstr=''):
        return memlink_cmd_insert(self.client, table, key, value, len(value), attrstr, pos)

    def sortlist_insert(self, table, key, value, attrstr=''):
        return memlink_cmd_sortlist_insert(self.client, table, key, value, len(value), attrstr)

    def sortlist_range(self, table, key, kind, valmin, valmax, attrstr=''):
        result = MemLinkResult()
        ret = memlink_cmd_sortlist_range(self.client, table, key, kind, attrstr, 
                                valmin, len(valmin), valmax, len(valmax), result)
        if ret != MEMLINK_OK:
            result = None
        return ret, result
    
    def sortlist_del(self, table, key, kind, valmin, valmax, attrstr=''):
        return memlink_cmd_sortlist_del(self.client, table, key, kind, attrstr, valmin, len(valmin), valmax, len(valmax))

    def sortlist_count(self, table, key, valmin, valmax, attrstr=''):
        result = MemLinkCount()
        ret = memlink_cmd_sortlist_count(self.client, table, key, attrstr, 
                                valmin, len(valmin), valmax, len(valmax), result)
        if ret != MEMLINK_OK:
            result = None
        return ret, result

    def move(self, table, key, value, pos):
        return memlink_cmd_move(self.client, table, key, value, len(value), pos)

    def attr(self, table, key, value, attrstr):
        return memlink_cmd_attr(self.client, table, key, value, len(value), attrstr)

    def tag(self, table, key, value, tag):
        return memlink_cmd_tag(self.client, table, key, value, len(value), tag)

    def range(self, table, key, kind, frompos, rlen, attrstr=''):
        result = MemLinkResult()
        ret = memlink_cmd_range(self.client, table, key, kind, attrstr, frompos, rlen, result)
        if ret != MEMLINK_OK:
            result = None
        return ret, result

    def rmkey(self, table, key):
        return memlink_cmd_rmkey(self.client, table, key)

    def count(self, table, key, attrstr=''):
        result = MemLinkCount()
        ret = memlink_cmd_count(self.client, table, key, attrstr, result)
        if ret != MEMLINK_OK:
            result = None
        return ret, result

    def lpush(self, table, key, value, attrstr=''):
        return memlink_cmd_lpush(self.client, table, key, value, len(value), attrstr)

    def rpush(self, table, key, value, attrstr=''):
        return memlink_cmd_rpush(self.client, table, key, value, len(value), attrstr)

    def lpop(self, table, key, num=1):
        result = MemLinkResult()
        ret = memlink_cmd_lpop(self.client, table, key, num, result)
        if ret != MEMLINK_OK:
            result = None
        return ret, result

    def rpop(self, table, key, num=1):
        result = MemLinkResult()
        ret = memlink_cmd_rpop(self.client, table, key, num, result)
        if ret != MEMLINK_OK:
            result = None
        return ret, result
    
    def read_conn_info(self):
        info = MemLinkRcInfo()
        ret = memlink_cmd_read_conn_info(self.client, info)
        if ret != MEMLINK_OK:
            info = None
        return ret, info

    def write_conn_info(self):
        info = MemLinkWcInfo()
        ret = memlink_cmd_write_conn_info(self.client, info)
        if ret != MEMLINK_OK:
            info = None
        return ret, info

    def sync_conn_info(self):
        info = MemLinkScInfo()
        ret = memlink_cmd_sync_conn_info(self.client, info)
        if ret != MEMLINK_OK:
            info = None
        return ret, info

    def insert_mkv(self, imkv):
        mkvobj = memlink_imkv_create()
        if imkv == []:
            return MEMLINK_ERR_PARAM, mkvobj
        for tup in imkv:
            if type(tup) != types.TupleType:
                return MEMLINK_ERR_PARAM, mkvobj
            if len(tup) != 4:
                return MEMLINK_ERR_PARAM, mkvobj 
            keyobj = memlink_ikey_create(tup[0], len(tup[0]))
            valobj = memlink_ival_create(str(tup[1]), len(tup[1]), str(tup[2]), int(tup[3]))
            ret = memlink_ikey_add_value(keyobj, valobj)
            ret = memlink_imkv_add_key(mkvobj, keyobj)

        ret = memlink_cmd_insert_mkv(self.client, mkvobj)
        return ret, mkvobj

def memlinkmkv_free(self):
    if MemLinkInsertMkv.is_free == 0:
        memlink_imkv_destroy(self)
        MemLinkInsertMkv.is_free = 1

MemLinkInsertMkv.close   = memlinkmkv_free
MemLinkInsertMkv.is_free = 0 
MemLinkInsertMkv.__del__ = memlinkmkv_free

def memlinkresult_list(self):
    v = []
    item = self.items
    while item:
        v.append([item.value, item.attr])
        item = item.next
    return v

def memlinkresult_strlist(self):
    v = []
    item = self.items
    while item:
        pos = item.value.find('\x00')
        if pos >= 0:
            v.append([item.value[:pos], item.attr])
        else:
            v.append([item.value, item.attr])
        item = item.next
    return v


def memlinkresult_free(self):
    memlink_result_free(self)

def memlinkresult_print(self):
    s = 'count:%d valuesize:%d attrsize:%d\n' % (self.count, self.valuesize, self.attrsize)
    item = self.items
    while item:
        s += 'value:%s attr:%s\n' % (item.value, item.attr)
        item = item.next
    return s

MemLinkResult.list    = memlinkresult_list
MemLinkResult.strlist = memlinkresult_strlist
MemLinkResult.close   = memlinkresult_free
MemLinkResult.__del__ = memlinkresult_free
MemLinkResult.__str__ = memlinkresult_print

def memlinkrcinfo_free(self):
    memlink_rcinfo_free(self)

def memlinkrcinfo_print(self):
    s = 'read count: %d\n' % (self.conncount)
    item = self.root
    while item:
        s += 'fd:%s ip:%s port:%s cmd:%s conn_time:%s\n' % \
                (item.fd, item.client_ip, item.port, item.cmd_count, item.conn_time)
        item = item.next
    return s

MemLinkRcInfo.close   = memlinkrcinfo_free
MemLinkRcInfo.__del__ = memlinkrcinfo_free
MemLinkRcInfo.__str__ = memlinkrcinfo_print

def memlinkwcinfo_free(self):
    memlink_wcinfo_free(self)

def memlinkwcinfo_print(self):
    s = 'write count: %d\n' % (self.conncount)
    item = self.root
    while item:
        s += 'fd:%s ip:%s port:%s cmd:%s conn_time:%s\n' % \
                (item.fd, item.client_ip, item.port, item.cmd_count, item.conn_time)
        item = item.next
    return s

MemLinkWcInfo.close   = memlinkwcinfo_free
MemLinkWcInfo.__del__ = memlinkwcinfo_free
MemLinkWcInfo.__str__ = memlinkwcinfo_print

def memlinkscinfo_free(self):
    memlink_scinfo_free(self)

def memlinkscinfo_print(self):
    s = 'slave count: %d\n' % (self.conncount)
    item = self.root
    while item:
        s += 'fd:%s ip:%s port:%s cmd:%s conn_time:%s logver:%s logline:%s delay:%s\n' % \
                (item.fd, item.client_ip, item.port, item.cmd_count, item.conn_time, item.logver, item.logline, item.delay)
        item = item.next
    return s

MemLinkScInfo.close   = memlinkscinfo_free
MemLinkScInfo.__del__ = memlinkscinfo_free
MemLinkScInfo.__str__ = memlinkscinfo_print


def memlinkstat_print(self):
    s = 'valuesize:%d\nattrsize:%d\nblocks:%d\ndata:%d\ndata_used:%d\nmem:%d\n' % \
        (self.valuesize, self.attrsize, self.blocks, self.data, self.data_used, self.mem)

    return s

MemLinkStat.__str__ = memlinkstat_print

def memlinkstatsys_print(self):
    s = 'keys:%d\nvalues:%d\nblocks:%d\ndata_all:%d\nht_mem:%d\npool.mem:%d\npool_blocks:%d\nall_mem:%d\nlogver:%d\nlogline:%d\n' % \
        (self.keys, self.values, self.blocks, self.data_all, self.ht_mem, self.pool_mem, self.pool_blocks, self.all_mem, self.logver, self.logline)

    return s

MemLinkStatSys.__str__ = memlinkstatsys_print

File: client/php/memlinkclient.php
<?php
include_once("cmemlink.php");

class MemLinkClient
{
    private $client;

    function __construct($host, $readport, $writeport, $timeout)
    {
        $this->client = memlink_create($host, $readport, $writeport, $timeout);
        //$r = memlink_create($host,$readport,$writeport,$timeout);
        //$this->client = is_resource($r) ? new MemLink($r) : $r;
    }

    function close()
    {
        if ($this->client) {
            memlink_close($this->client);
        }
    }

    function destroy()
    {
        if ($this->client) {
            memlink_destroy($this->client);
        }
        $this->client = null;
    }

	function create_table($table, $valuesize, $attrformat, $listtype, $valuetype)
	{
    	if( false == is_int($valuesize) or false == is_string($table) or false == is_string($attrformat) )
    	{
    		return -1;
    	}
    	
		return memlink_cmd_create($this->client, $table, $valuesize, $attrformat, $listtype, $valuetype);
	}

	function create_table_list($table, $valuesize, $attrformat)
	{
    	if( false == is_int($valuesize) or false == is_string($table) or false == is_string($attrformat) )
    	{
    		return -1;
    	}
    	
		return memlink_cmd_create_list($this->client, $table, $valuesize, $attrformat);
	}

	function create_table_queue($table, $valuesize, $attrformat)
	{
    	if( false == is_int($valuesize) or false == is_string($table) or false == is_string($attrformat) )
    	{
    		return -1;
    	}
    	
		return memlink_cmd_create_queue($this->client, $table, $valuesize, $attrformat);
	}

    function create_table_sortlist($table, $valuesize, $attrformat, $valuetype)
	{
    	if( false == is_int($valuesize) or false == is_string($table) or false == is_string($attrformat) )
    	{
    		return -1;
    	}
    	
		return memlink_cmd_create_sortlist($this->client, $table, $valuesize, $attrformat, $valuetype);
	}
	
	function remove_table($table)
	{
		if (false == is_string($table)) {
			return -1;
		}
		return memlink_cmd_remove_table($this->client, $table);
	}

	
	function create_node($table, $key)
	{
		if (false == is_string($table) or false == is_string($key)) {
			return -1;
		}
		return memlink_cmd_create_node($this->client, $table, $key);
	}

	function ping()
    {
        return memlink_cmd_ping($this->client);
    }

    function dump()
    {
        return memlink_cmd_dump($this->client);
    }

    function clean($table, $key)
    {
    	if(false == is_string($table) || false == is_string($key)) {
    		return -1;
    	}
        return memlink_cmd_clean($this->client, $table, $key);
    }

    function stat($table, $key)
    {
    	if( false == is_string($table) || false == is_string($key))
    	{
    		return NULL;
    	}
    
        $mstat = new MemLinkStat();

        $ret = memlink_cmd_stat($this->client, $table, $key, $mstat);
		if ($ret == MEMLINK_OK) {
			return $mstat;
		}
		return NULL;
    }

	function stat2($table, $key, $mstat)
	{
    	if( false == is_string($table) || false == is_string($key)) {
    		return -1;
    	}
	
        return memlink_cmd_stat($this->client, $table, $key, $mstat);
	}

    function stat_sys()
    {
        $stat = new MemLinkStatSys();
        $ret = memlink_cmd_stat_sys($this->client, $stat);
        if ($ret == MEMLINK_OK) {
            return $stat;
        }
        return NULL;
    }

    function stat_sys2($stat)
    {
        return memlink_cmd_stat_sys($this->client, $stat);
    }

    function delete($table, $key, $value, $valuelen)
    {
    	if( false == is_string($table) || false == is_string($key) or false == is_int($valuelen) )
    	{
    		return -1;
    	}
    	
        return memlink_cmd_del($this->client, $table, $key, $value, $valuelen);
    }

    function insert($table, $key, $value, $valuelen, $attrstr, $pos)
    {
    	if( false == is_string($table) || false == is_string($key) or false == is_int($valuelen) or
    		false == is_int($pos) or false == is_string($attrstr) )
    	{
    		return -1;
    	}
		
        return memlink_cmd_insert($this->client, $table, $key, $value, $valuelen, $attrstr, $pos);
    }

    function move($table, $key, $value, $valuelen, $pos)
    {
    	if( false == is_string($table) || false == is_string($key) or false == is_int($valuelen) or false == is_int($pos) )
    	{
    		return -1;
    	}
			
        return memlink_cmd_move($this->client, $table, $key, $value, $valuelen, $pos);
    }

    function attr($table, $key, $value, $valuelen, $attrstr)
    {
    	if( false == is_string($table) || false == is_string($key) or false == is_int($valuelen) or false == is_string($attrstr) )
    	{
    		return -1;
    	}
    
        return memlink_cmd_attr($this->client, $table, $key, $value, $valuelen, $attrstr);
    }

    function tag($table, $key, $value, $valuelen, $tag)
    {
    	if( false == is_string($table) || false == is_string($key) or false == is_int($valuelen) or false == is_int($tag) )
    	{
    		return -1;
    	}
    	
        return memlink_cmd_tag($this->client, $table, $key, $value, $valuelen, $tag);
    }

    function range($table, $key, $kind, $attrstr, $frompos, $len)
    {
    	if( false == is_string($table) || false == is_string($key) or false == is_int($frompos) or
    		false == is_int($len) or false == is_string($attrstr) )
    	{
    		return NULL;
    	}
    	
        $result = new MemLinkResult();
        
        $ret = memlink_cmd_range($this->client, $table, $key, $kind, $attrstr, $frompos, $len, $result);
		if ($ret == MEMLINK_OK) {
			return $result;
		}
		return NULL;
    }

	function range_visible($table, $key, $attrstr, $frompos, $len) 
	{
		return $this->range($table, $key, MEMLINK_VALUE_VISIBLE, $attrstr, $frompos, $len);
	}
	
	function range_tagdel($table, $key, $attrstr, $frompos, $len) 
	{
		return $this->range($table, $key, MEMLINK_VALUE_TAGDEL, $attrstr, $frompos, $len);
	}

	function range_all($table, $key, $attrstr, $frompos, $len) 
	{
		return $this->range($table, $key, MEMLINK_VALUE_ALL, $attrstr, $frompos, $len);
	}

	function range2($table, $key, $kind, $attrstr, $frompos, $len, $result)
	{
    	if( false == is_string($table) || false == is_string($key) or false == is_int($frompos) or false == is_int($len) or false == is_string($attrstr) )
    	{
    		return -1;
    	}
	
		return memlink_cmd_range($this->client, $table, $key, $kind, $attrstr, $frompos, $len, $result);
	}

	function range2_visible($table, $key, $attrstr, $frompos, $len, $result)
	{
		return $this->range2($table, $key, MEMLINK_VALUIE_VISIBLE, $attrstr, $frompos, $len, $result);
	}

	function range2_tagdel($table, $key, $attrstr, $frompos, $len, $result)
	{
		return $this->range2($table, $key, MEMLINK_VALUIE_TAGDEL, $attrstr, $frompos, $len, $result);
	}

	function range2_all($table, $key, $attrstr, $frompos, $len, $result)
	{
		return $this->range2($table, $key, MEMLINK_VALUIE_ALL, $attrstr, $frompos, $len, $result);
	}

	function rmkey($table, $key)
	{
    	if( false == is_string($table) || false == is_string($key) ) {
    		return -1;
    	}
	
		return memlink_cmd_rmkey($this->client, $table, $key);
	}

	function count($table, $key, $attrstr)
	{
    	if( false == $table || false == is_string($key) or false == is_string($attrstr) ) {
    		return NULL;
    	}
	
		$count = new MemLinkCount();
		$ret = memlink_cmd_count($this->client, $table, $key, $attrstr, $count);
		if ($ret == MEMLINK_OK) {
			return $count;
		}
		return NULL;
	}

	function count2($table, $key, $attrstr, $count)
	{
    	if( false == is_string($table) || false == is_string($key) or false == is_string($attrstr) ) {
    		return -1;
    	}
	
		return memlink_cmd_count($this->client, $table, $key, $attrstr, $count);
	}

    function lpush($table, $key, $value, $valuelen, $attrstr="")
    {
    	if( false == is_string($table) || false == is_string($key) or false == is_string($attrstr) ) {
    		return -1;
    	}
	
        return memlink_cmd_lpush($this->client, $table, $key, $value, $valuelen, $attrstr);
    }

    function rpush($table, $key, $value, $valuelen, $attrstr="")
    {
    	if( false == is_string($table) || false == is_string($key) or false == is_string($attrstr) ) {
    		return -1;
    	}
	
        return memlink_cmd_rpush($this->client, $table, $key, $value, $valuelen, $attrstr);
    }

    function lpop($table, $key, $num=1)
    {
        $result = new MemLinkResult();
        $ret = memlink_cmd_lpop($this->client, $table, $key, $num, $result);
		if ($ret == MEMLINK_OK) {
			return $result;
		}
		return NULL;
    }

    function lpop2($table, $key, $num, $result)
    {
        return memlink_cmd_lpop($this->client, $table, $key, $num, $result);
    }


    function rpop($table, $key, $num=1)
    {
        $result = new MemLinkResult();
        $ret = memlink_cmd_lpop($this->client, $table, $key, $num, $result);
		if ($ret == MEMLINK_OK) {
			return $result;
		}
		return NULL;
    }
    
    function rpop2($table, $key, $num, $result)
    {
        return memlink_cmd_rpop($this->client, $table, $key, $num, $result);
    }
    
    function sortlist_insert($table, $key, $value, $valuelen, $attrstr="")
    {
        return memlink_cmd_sortlist_insert($this->client, $table, $key, $value, $valuelen, $attrstr);
    }

    function sortlist_range($table, $key, $kind, $vmin, $vminlen, $vmax, $vmaxlen, $attrstr='')
    {
        $result = new MemLinkResult();
        $ret = memlink_cmd_sortlist_range($this->client, $table, $key, $kind, $attrstr, $vmin, $vminlen, $vmax, $vmaxlen, $result);
		if ($ret == MEMLINK_OK) {
			return $result;
		}
		return NULL;
    }

    function sortlist_range2($table, $key, $kind, $vmin, $vminlen, $vmax, $vmaxlen, $attrstr, $result)
    {
        return memlink_cmd_sortlist_range($this->client, $table, $key, $kind, $attrstr, $vmin, $vminlen, $vmax, $vmaxlen, $result);
    }

    function sortlist_del($table, $key, $kind, $vmin, $vminlen, $vmax, $vmaxlen, $attrstr='')
    {
        return memlink_cmd_sortlist_del($this->client, $table, $key, $kind, $attrstr, $vmin, $vminlen, $vmax, $vmaxlen);
    }

    function sortlist_count($table, $key, $vmin, $vminlen, $vmax, $vmaxlen, $attrstr='')
    {
        $result = new MemLinkCount();
        $ret = memlink_cmd_sortlist_count($this->client, $table, $key, $attrstr, $vmin, $vminlen, $vmax, $vmaxlen, $result);
        if ($ret == MEMLINK_OK) {
            return $result;
        }
        return NULL;
    }

    function sortlist_count2($table, $key, $vmin, $vminlen, $vmax, $vmaxlen, $attrstr, $result)
    {
        return memlink_cmd_sortlist_count($this->client, $table, $key, $attrstr, $vmin, $vminlen, $vmax, $vmaxlen, $result);
    }

    function read_conn_info()
    {
        $rcinfo = new MemLinkRcInfo();
        $ret = memlink_cmd_read_conn_info($this->client, $rcinfo);
        if ($ret == MEMLINK_OK) {
            return $rcinfo;
        }
        return NULL;
    }
    function write_conn_info()
    {
        $wcinfo = new MemLinkWcInfo();
        $ret = memlink_cmd_write_conn_info($this->client, $wcinfo);
        if ($ret == MEMLINK_OK) {
            return $wcinfo;
        }
        return NULL;
    }
    function sync_conn_info()
    {
        $scinfo = new MemLinkScInfo();
        $ret = memlink_cmd_sync_conn_info($this->client, $scinfo);
        if ($ret == MEMLINK_OK) {
            return $scinfo;
        }
        return NULL;
    }
    static function memlink_imkv_create() {
        $r=memlink_imkv_create();
        if (is_resource($r)) {
            $c=substr(get_resource_type($r), (strpos(get_resource_type($r), '__') ? strpos(get_resource_type($r), '__') + 2 : 3));
            if (!class_exists($c)) {
                return new MemLinkInsertMkv($r);
            }
            return new $c($r);
        }
        return $r;
    }
    static function memlink_ikey_create($table, $key,$keylen) {
        $r=memlink_ikey_create($table, $key,$keylen);
        if (is_resource($r)) {
            $c=substr(get_resource_type($r), (strpos(get_resource_type($r), '__') ? strpos(get_resource_type($r), '__') + 2 : 3));
            if (!class_exists($c)) {
                return new MemLinkInsertKey($r);
            }
            return new $c($r);
        }
        return $r;
    }
    static function memlink_ival_create($value,$valuelen,$attrstr,$pos) {
        $r=memlink_ival_create($value,$valuelen,$attrstr,$pos);
        if (is_resource($r)) {
            $c=substr(get_resource_type($r), (strpos(get_resource_type($r), '__') ? strpos(get_resource_type($r), '__') + 2 : 3));
            if (!class_exists($c)) {
                return new MemLinkInsertVal($r);
            }
            return new $c($r);
        }
        return $r;
    }

    function insert_mkv($array)
    {
        $mkv = $this->memlink_imkv_create();
        foreach ($array as $item) {
            if (is_array($item) && count($item) == 4) {
                if (is_null($item["key"]) || is_null($item["value"]) || 
                    is_null($item["attr"]) || is_null($item["pos"])) {
                    return $mkv;
                }
                $keyobj = $this->memlink_ikey_create($item["key"], strlen($item["key"]));
                $valobj = $this->memlink_ival_create($item["value"], strlen($item["value"]), $item["attr"], $item["pos"]);
                $ret = memlink_ikey_add_value($keyobj, $valobj); 
                if ($ret != MEMLINK_OK) {
                    return $mkv;
                }
                $ret = memlink_imkv_add_key($mkv, $keyobj);
                if ($ret != MEMLINK_OK) {
                    return $mkv;
                }
            }
        }
        $ret = memlink_cmd_insert_mkv($this->client, $mkv);
        return $mkv;
    }

    function rcinfo_free($info) {
        return memlink_rcinfo_free($info);
    }

    function wcinfo_free($info) {
        return memlink_wcinfo_free($info);
    }

    function scinfo_free($info) {
        return memlink_scinfo_free($info);
    }

    function mkv_destroy($mkv) {
        return memlink_imkv_destroy($mkv);
    }
}

?>
